src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp

```diff
@@ -207,81 +287,162 @@ void VehicleAngularVelocity::Run()
+						_angular_velocity_prev = _angular_velocity / sensor_fifo_data.scale;
+						ResetFilters(_angular_velocity / sensor_fifo_data.scale, _angular_acceleration / sensor_fifo_data.scale);
 
-		// Guard against too small (< 0.2ms) and too large (> 20ms) dt's.
-		const float dt = math::constrain(((sensor_data.timestamp_sample - _timestamp_sample_prev) / 1e6f), 0.0002f, 0.02f);
-		_timestamp_sample_prev = sensor_data.timestamp_sample;
+						_fifo_last_scale = sensor_fifo_data.scale;
+					}
+
+					if (_reset_filters) {
+						continue; // not safe to run until filters configured
+					}
+				}
+
+				Vector3f angular_velocity_unscaled;
+				Vector3f angular_acceleration_unscaled;
 
-		// get the sensor data and correct for thermal errors (apply offsets and scale)
-		const Vector3f val{sensor_data.x, sensor_data.y, sensor_data.z};
+				int16_t *raw_data_array[] {sensor_fifo_data.x, sensor_fifo_data.y, sensor_fifo_data.z};
 
-		// correct for in-run bias errors
-		const Vector3f angular_velocity_raw = _calibration.Correct(val) - _bias;
+				for (int axis = 0; axis < 3; axis++) {
+					// copy raw int16 sensor samples to float array for filtering
+					float data[FIFO_SIZE_MAX];
 
-		// Gyro filtering:
-		// - Apply general notch filter (IMU_GYRO_NF_FREQ)
-		// - Apply general low-pass filter (IMU_GYRO_CUTOFF)
-		// - Differentiate & apply specific angular acceleration (D-term) low-pass (IMU_DGYRO_CUTOFF)
+					for (int n = 0; n < N; n++) {
+						data[n] = raw_data_array[axis][n];
+					}
 
-		const Vector3f angular_velocity_notched{_notch_filter_velocity.apply(angular_velocity_raw)};
+					// Apply general notch filter (IMU_GYRO_NF_FREQ)
+					if (_notch_filter_velocity[axis].getNotchFreq() > 0.f) {
+						_notch_filter_velocity[axis].apply(data, N);
+					}
 
-		const Vector3f angular_velocity{_lp_filter_velocity.apply(angular_velocity_notched)};
+					// Apply general low-pass filter (IMU_GYRO_CUTOFF)
+					_lp_filter_velocity[axis].apply(data, N);
 
-		const Vector3f angular_acceleration_raw = (angular_velocity - _angular_velocity_prev) / dt;
-		_angular_velocity_prev = angular_velocity;
-		_angular_acceleration_prev = angular_acceleration_raw;
-		const Vector3f angular_acceleration{_lp_filter_acceleration.apply(angular_acceleration_raw)};
+					// save last filtered sample
+					angular_velocity_unscaled(axis) = data[N - 1];
 
 
-		// publish once all new samples are processed
-		if (!_sensor_sub.updated()) {
-			bool publish = true;
+					// angular acceleration: Differentiate & apply specific angular acceleration (D-term) low-pass (IMU_DGYRO_CUTOFF)
+					float delta_velocity_filtered;
 
-			if (_param_imu_gyro_rate_max.get() > 0) {
-				const uint64_t interval = 1e6f / _param_imu_gyro_rate_max.get();
+					for (int n = 0; n < N; n++) {
+						const float delta_velocity = (data[n] - _angular_velocity_prev(axis));
+						delta_velocity_filtered = _lp_filter_acceleration[axis].apply(delta_velocity);
```

Comment:
Is that better than creating an array of "delta velocities" and then sending it to the `Array` version of the filter?

Reply:
No, I only noticed I could move the dt out of the loop yesterday. Let's get this in and I'll do that in a followup.